src/shared/components/chartjs/plot.tsx

```diff
@@ -0,0 +1,122 @@
+import React, { MouseEvent, useEffect, useMemo, useRef, useState } from "react";
+import Chart, { ChartData, ChartOptions } from "chart.js";
+import "chartjs-plugin-dragdata";
+import { usePrevious } from "../../hooks/use-previous";
+import { pixelsToCoords, nearestPoint } from "./calculate-coordinates";
+
+interface PropsT {
+  externalRefresh?: boolean; // allows parent to mutate data and yield refresh
+  data: ChartData;
+  options: ChartOptions;
+  dragOptions: DragOptionsT;
+  interactive: boolean;
+  onDoubleClickCoords?: (x: Number, y: Number) => void;
+}
+
+interface DragOptionsT {
+  onDrag?: (e: any, datasetIndex: number, index: number, value: number) => void;
+  onDragEnd?: (
+    e: any,
+    datasetIndex: number,
+    index: number,
+    value: number
+  ) => void;
+}
+
+export default (props: PropsT) => {
+  const {
+    data,
+    options,
+    dragOptions,
+    interactive,
+    onDoubleClickCoords,
+    externalRefresh
+  } = props;
+
+  const [internalRefresh, setInternalRefresh] = useState(false);
+
+  const refreshSwitch =
+    usePrevious(false) != (externalRefresh || internalRefresh);
+
+  console.log("RefreshSwitch: " + refreshSwitch);
+
+  const canvasRef = useRef(null);
+  const chartRef = useRef<Chart | null>(null);
+
+  const fullOptions = useMemo(
+    () => ({
+      ...options,
+      ...(interactive ? dragOptions : {}),
+      dragData: interactive
+    }),
+    [options, dragOptions, interactive, refreshSwitch]
+  );
+
+  const modifiedData = useMemo(() => {
+    return {
+      datasets: (data.datasets || []).map(ds => {
+        return {
+          ...ds,
+          showLine: true
+        };
+      }),
+      ...data
+    };
+  }, [data, refreshSwitch]);
+
+  useEffect(() => {
+    console.log("Using effect");
+    if (!canvasRef.current) {
+      return;
+    }
+    // @ts-ignore
+    const ctx = canvasRef.current.getContext("2d");
+
+    if (!chartRef.current) {
+      chartRef.current = new Chart(ctx, {
+        type: "scatter",
+        data,
+        options: fullOptions
+      });
+    } else {
+      if (externalRefresh) {
+        chartRef.current.options = fullOptions;
+        chartRef.current.data = data;
+      }
+    }
+    chartRef.current.update();
+
+    if (internalRefresh) {
+      setInternalRefresh(false);
+    }
+  });
+
+  // cleanup by destroying chart
+  useEffect(
+    () => () => {
+      chartRef.current && chartRef.current.destroy();
+    },
+    []
+  );
```

Comment:
Compiler warning: React Hook useEffect contains a call to 'setInternalRefresh'. Without a list of dependencies, this can lead to an infinite chain of updates. To fix this, pass [internalRefresh, data, fullOptions, externalRefresh] as a second argument to the useEffect Hook  react-hooks/exhaustive-deps

Reply:
Ignore it. We're using effects in an "unsafe way" due to the mutative nature of the implementation. `setInternalRefresh(false)` will ensure that this branch can't be taken twice.